packages/mobile/src/services/offline-downloader/offline-sync.ts

```diff
@@ -0,0 +1,253 @@
+      }
+    })
+  )
+  batchRemoveTrackDownload(tracksForDelete)
+}
+
+const syncCollection = async (
+  offlineCollection: Collection,
+  isFavoritesDownload?: boolean
+) => {
+  // TODO: record and check last verified time for collections
+  const state = store.getState()
+  const currentUserId = getUserId(state as unknown as CommonState)
+  const collectionId = offlineCollection.playlist_id
+  const collectionIdStr = offlineCollection.playlist_id.toString()
+  const updatedCollection: UserCollectionMetadata | undefined = (
+    await apiClient.getPlaylist({
+      playlistId: collectionId,
+      currentUserId
+    })
+  )?.[0]
+
+  // TODO: will discovery serve a removed playlist?
+  if (!updatedCollection) return
+
+  if (
+    moment(updatedCollection.updated_at).isSameOrBefore(
+      offlineCollection.updated_at
+    )
+  ) {
+    return
+  }
+
+  const updatedCollectionWithArt = await populateCoverArtSizes(
+    updatedCollection
+  )
+
+  downloadCollection(updatedCollectionWithArt)
+  if (
+    updatedCollectionWithArt.cover_art_sizes !==
+    offlineCollection.cover_art_sizes
+  ) {
+    downloadCollectionCoverArt(updatedCollectionWithArt)
+  }
+
+  const oldTrackIds = new Set(
+    offlineCollection.tracks?.map((track) => track.track_id)
+  )
+  const newTrackIds = new Set(
+    updatedCollection.tracks?.map((track) => track.track_id)
+  )
+  const addedTrackIds = [...newTrackIds].filter(
+    (trackId) => !oldTrackIds.has(trackId)
+  )
+  const removedTrackIds = [...oldTrackIds].filter(
+    (trackId) => !newTrackIds.has(trackId)
+  )
+
+  const tracksForDelete: TrackForDownload[] = removedTrackIds.map(
+    (removedTrack) => ({
+      trackId: removedTrack,
+      downloadReason: {
+        is_from_favorites: isFavoritesDownload,
+        collection_id: collectionIdStr
+      }
+    })
+  )
+  removeCollectionDownload(collectionIdStr, tracksForDelete)
+
+  const tracksForDownload: TrackForDownload[] = addedTrackIds.map(
+    (addedTrack) => ({
+      trackId: addedTrack,
+      downloadReason: {
+        is_from_favorites: isFavoritesDownload,
+        collection_id: collectionIdStr
+      }
+    })
+  )
+  batchDownloadTrack(tracksForDownload)
+}
+
+const syncStaleTracks = () => {
+  const state = store.getState()
+  const cacheTracks = getTracks(state, {})
+  const currentUserId = getUserId(state as unknown as CommonState)
+  if (!currentUserId) return
+
+  const staleCachedTracks = Object.entries(cacheTracks)
+    .filter(
+      ([id, track]) =>
+        track.offline &&
+        moment()
+          .subtract(STALE_DURATION_TRACKS)
+          .isAfter(moment(track.offline?.last_verified_time))
+    )
+    .map(([id, track]) => track)
+
+  staleCachedTracks.forEach(async (staleTrack) => {
+    const updatedTrack: UserTrackMetadata = await apiClient.getTrack({
+      id: staleTrack.track_id,
+      currentUserId
+    })
+
+    // If track should not be available
+    if (
+      !updatedTrack.is_available ||
```

Comment:
nice this is a great block, i wonder if we use this in other spots around the app? might be a nice helper just cause i bet it's hard as a new dev to know what qualifies as "not playable'

Reply:
Totally. I'm not even 100% that I got them all or that all of these should be blocked. I can move it to a utility